### PR TITLE
Chat avatars

### LINKED ISSUE #179: Chat should show user avatars
As I understand it, you can access the user avatar via user.avatar. This would make it possible for the chat to show the user avatar instead of a placeholder. For dynamic messages this is harder, a path like users/id/avatar would be nice if it doesn't exist yet.

- [x] Chat that is rendered on reload shows avatars
- [x] Chat that is generated via web socket shows avatars

By the way, the user list on the dashboard should show avatars as well

### LINKED ISSUE #185: Integrate Profile Pictures in global chat
The Icon on the side of sent messages should be replaced with the profile picture of the user that sent the message

### PR DESCRIPTION
**Pull request by konradh**

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -1,30 +1,28 @@
...
+  end
+
+  def filter_users_status
+    @users = User.where.not(id: current_user.id).with_status(User.filter_status)
+    @users = @users.sample(maximum_length_user_list)
+  end
+
   def maximum_length_user_list
     30
   end
```
Comment (frcroth) [inline app/controllers/home_controller.rb]: This will result in a merge conflict. But you should be able to handle it. maximum_length_user_list is a field now.

File: app/views/home/_chat.html.erb [hunk 2]
```diff
@@ -19,11 +19,7 @@
 <div class="d-none" data-role="message-template">
   <div class="chat-item d-flex flex-row">
     <div class="avatar-container d-flex align-items-end">
-      <a href="#">
-        <span class="w-48 avatar">
-          <i class="fa fa-user"></i>
-        </span>
-      </a>
+      <%= render 'users/avatar', user: current_user, size: 48, classes: %i[rounded-circle] %>
```
Comment (frcroth) [inline app/views/home/_chat.html.erb]: Is rounded-circle necessary here? Shouldn't all occurrences of avatar be circles?

File: spec/features/global_chat_spec.rb [hunk 3]
```diff
@@ -35,17 +30,29 @@ def post_message(message)
...
+    end
+
+    it "displays the user's avatar" do
+      expect(page).to have_css("img[src='#{avatar_user_path(user)}']")
+    end
+
+    it 'shows the correct timestamp after reload' do
+      visit root_path
+      expect(page).to have_text(room_message.formatted_time)
+    end
```
Comment (frcroth) [inline spec/features/global_chat_spec.rb:68]: Add a test that it links to the user profile

File: spec/features/global_chat_spec.rb [hunk 4]
```diff
@@ -35,17 +30,29 @@ def post_message(message)
...
+    it 'clears text box after message submit' do
+      expect(page).to have_field('room_message_message', text: '')
+    end
+
+    it 'shows the correct timestamp' do
+      expect(page).to have_text(room_message.formatted_time)
+    end
+
+    it "displays the user's avatar" do
+      expect(page).to have_css("img[src='#{avatar_user_path(user)}']")
```
Comment (frcroth) [inline spec/features/global_chat_spec.rb]: Use `within('#chat-messages')` to make it more specific

**Review by frcroth — CHANGES_REQUESTED** [review]

File: app/views/home/_chat.html.erb (hunk 2 above)
Comment (felixauringer) [reply app/views/home/_chat.html.erb]: If that is the case, it can be moved to the avatar partial

File: spec/features/global_chat_spec.rb [hunk 5]
```diff
@@ -35,17 +30,29 @@ def post_message(message)
     end
   end
 
-  it 'shows the correct timestamp' do
-    post_message(message)
-    message = RoomMessage.all.last
-    expect(page).to have_text(message.formatted_time)
-  end
+  context 'with a posted message' do
+    let(:room_message) { RoomMessage.all.last }
```
Comment (felixauringer) [inline spec/features/global_chat_spec.rb]: Why do we need this extra variable? Shouldn't the last message equal the outer `let(:message)`?
And if this is to check whether the message was saved, then I would add a new test for that rather than checking it implicitly by getting the last message from the database

File: spec/features/global_chat_spec.rb (hunk 4 above)
Comment (felixauringer) [reply spec/features/global_chat_spec.rb]: Could be added to some more tests here

File: spec/features/global_chat_spec.rb [hunk 6]
```diff
@@ -35,17 +30,29 @@ def post_message(message)
...
+
+    it 'shows the correct timestamp' do
+      expect(page).to have_text(room_message.formatted_time)
+    end
+
+    it "displays the user's avatar" do
+      expect(page).to have_css("img[src='#{avatar_user_path(user)}']")
+    end
+
+    it 'shows the correct timestamp after reload' do
```
Comment (felixauringer) [inline spec/features/global_chat_spec.rb:63]: I would check for all things (message, time, avatar, avatar link, name and name link) after posting and after reloading. Should be easy using shared examples and only a different before block. Can also be moved to another PR if you would prefer that

**Review by felixauringer — CHANGES_REQUESTED** [review]

File: spec/features/global_chat_spec.rb (hunk 5 above)
Comment (konradh) [reply spec/features/global_chat_spec.rb]: The outer message is a String, but in this tests we need to call `formatted_time` on a RoomMessage object.

Comment (felixauringer) [reply spec/features/global_chat_spec.rb]: Hmm, I see. Confusing names here, maybe one could name the outer one `message_text`, `message_string` or something like that?

File: spec/features/global_chat_spec.rb [hunk 7]
```diff
@@ -35,17 +30,42 @@ def post_message(message)
...
+    let(:room_message) { RoomMessage.all.last }
 
-  it 'shows the correct timestamp after reload' do
-    post_message(message)
-    message = RoomMessage.all.last
-    visit root_path
-    expect(page).to have_text(message.formatted_time)
+    before do
+      post_message(message)
+      room_message
```
Comment (felixauringer) [inline spec/features/global_chat_spec.rb]: `let!` would do that for you. Then the before block could be a one liner with `{}`
EDIT: nope, it wouldn't because it is executed before :sweat_smile: Sorry

File: spec/features/global_chat_spec.rb [hunk 8]
```diff
@@ -35,17 +30,42 @@ def post_message(message)
...
+      expect(page).to have_field('room_message_message', text: '')
+    end
+
+    it 'shows the correct timestamp' do
+      within('#chat-messages') do
+        expect(page).to have_text(room_message.formatted_time)
+      end
+    end
+
+    it 'links to the users profile page' do
```
Comment (felixauringer) [inline spec/features/global_chat_spec.rb]: We would need two more specific tests for that because the user name (do we test if that is only displayed for messages from other users?) and the profile picture both have the same link.

With so many changes to the tests, we could really open a new PR (if you did not start with the tests yet)

File: app/views/users/_avatar.html.erb [hunk 9]
```diff
@@ -1,2 +1,4 @@
 <%= stylesheet_link_tag 'users', media: 'all', 'data-turbolinks-track': 'reload' %>
-<%= image_tag avatar_user_path(user), id: "avatar-#{user.id}", height: size, width: size, class: "#{classes.join(' ')}" %>
+<%= link_to user_path(user), class: "avatar-user-link" do %>
+  <%= image_tag avatar_user_path(user), id: "avatar-#{user.id}", height: size, width: size, class: "#{classes.append('avatar-image', 'rounded-circle', 'centered-avatar').join(' ')}" %>
```
Comment (felixauringer) [inline app/views/users/_avatar.html.erb]: With all these classes now in here, it would be possible to remove `classes` completely and just set it statically, wouldn't it?

**Review by felixauringer — APPROVED** [review]
I'm happy with the amount of tests now. There can always be more but I think it is fine now :+1:

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (felixauringer) [reply app/controllers/home_controller.rb]: This is resolved properly (by you, so I just resolve this :sweat_smile: )

**Review by frcroth — APPROVED** [review]
Merge it!
